tests/sentry/api/endpoints/test_organization_metric_data.py

```diff
@@ -1978,6 +1978,111 @@ def test_failure_count_transaction(self):
+                        ),
+                    },
+                    "type": "d",
+                    "value": [0.3],
+                    "retention_days": 90,
+                },
+                {
+                    "org_id": self.organization.id,
+                    "project_id": self.project.id,
+                    "metric_id": self.tx_metric,
+                    "timestamp": user_ts,
+                    "tags": {
+                        self.tx_status: indexer.record(
+                            self.organization.id, TransactionStatusTagValue.UNKNOWN.value
+                        ),
+                    },
+                    "type": "d",
+                    "value": [2.3],
+                    "retention_days": 90,
+                },
+                {
+                    "org_id": self.organization.id,
+                    "project_id": self.project.id,
+                    "metric_id": self.tx_metric,
+                    "timestamp": user_ts,
+                    "tags": {
+                        self.tx_status: indexer.record(
+                            self.organization.id, TransactionStatusTagValue.ABORTED.value
+                        ),
+                    },
+                    "type": "d",
+                    "value": [0.5],
+                    "retention_days": 90,
+                },
+            ],
+            entity="metrics_distributions",
+        )
+        response = self.get_success_response(
+            self.organization.slug,
+            field=["transaction.failure_rate"],
+            statsPeriod="1m",
+            interval="1m",
+        )
+
+        assert len(response.data["groups"]) == 1
+        group = response.data["groups"][0]
+        assert group["by"] == {}
+        assert group["totals"] == {"transaction.failure_rate": 25.0}
+        assert group["series"] == {"transaction.failure_rate": [25.0]}
+
+    def test_failure_rate_without_transactions(self):
+        """
+        Ensures the absence of transactions isn't an issue to calculate the rate.
+
+        The `nan` a division by 0 may produce must not be in the response, yet
+        they are an issue in javascript:
+        ```
+        $ node
+        Welcome to Node.js v16.13.1.
+        Type ".help" for more information.
+        > JSON.parse('NaN')
+        Uncaught SyntaxError: Unexpected token N in JSON at position 0
+        > JSON.parse('nan')
+        Uncaught SyntaxError: Unexpected token a in JSON at position 1
+        ```
+        """
+        # Not sending buckets means no project is created automatically. We need
+        # a project without transaction data, so create one:
+        self.project
+
+        response = self.get_success_response(
+            self.organization.slug,
+            field=["transaction.failure_rate"],
+            statsPeriod="1m",
+            interval="1m",
+        )
+
+        assert response.data["groups"] == []
```

Comment:
Do I understand correctly that the division by zero will never happen, because the snuba result will be empty instead? Or is there a case where `NaN`, `-Inf` or `Inf` pop up? In the second case, I would explicitly map them to `None`, as we do here: https://github.com/getsentry/sentry/blob/6fa42d944675f971b3e8ab8d5ae35d6d9740a1b6/src/sentry/snuba/sessions_v2.py#L140-L143

Reply:
`nan`s happen when explicitly doing `0/0`s on clickhouse (`divide(0, 0)`, actually); and I haven't investigated where and why, but at some point a `nan` is converted to `None`. I haven't investigated either what is going on when there isn't any data (the test we're looking at), but the groups end up being empty.

Since I haven't been able to reproduce the problematic case where `nan`s are present in the response, and I believe empty groups represent well the absence of data, I've assumed it's a good solution. What do you think? Should we be more explicit and set `None` when there's no data?

Reply:
> Since I haven't been able to reproduce the problematic case where nans are present in the response, and I believe empty groups represent well the absence of data, I've assumed it's a good solution.

Yeah sounds good, if an invalid float slips through after all, we will get notified in sentry.